tests/integrationv2/test_external_psk.py

```diff
@@ -0,0 +1,380 @@
+        mode=mode,
+        extra_flags=client_psk_params)
+    if certificate:
+        options.key = certificate.key
+        options.cert = certificate.cert
+        options.trust_store = certificate.cert
+    return options
+
+
+def get_psk_hash_alg_from_cipher(cipher):
+    # S2N supports only SHA256 and SHA384 PSK Hash Algorithms
+    if 'SHA256' in cipher.name:
+        return 'SHA256'
+    elif 'SHA384' in cipher.name: 
+        return 'SHA384'
+    else:
+        return None
+
+   
+def skip_invalid_psk_tests(provider, psk_hash_alg):
+    # If the PSK hash algorithm is None, it is not supported and we can safely skip the test case. 
+    if psk_hash_alg is None:
+        pytest.skip()
+
+    # In OpenSSL, PSK works only with TLS1.3 ciphersuites based on SHA256 hash algorithm which includes 
+    # all TLS1.3 ciphersuites supported by S2N except TLS_AES_256_GCM_SHA384.
+    if provider == OpenSSL and psk_hash_alg == 'SHA384':
+        pytest.skip()
+
+
+def validate_negotiated_psk_s2n(outcome, psk_identity, results):
+    if outcome == Outcome.psk_connection:
+        assert to_bytes("Negotiated PSK identity: {}".format(psk_identity)) in results.stdout
+    elif outcome == Outcome.full_handshake:
+        assert to_bytes("Negotiated PSK identity: {}".format(psk_identity)) not in results.stdout
+    else:
+        assert results.exit_code != 0
+        assert to_bytes("Failed to negotiate: 'TLS alert received'") in results.stderr
+
+
+def validate_negotiated_psk_openssl(outcome, results):
+    if outcome == Outcome.psk_connection:
+        assert to_bytes("extension \"psk\"") in results.stdout
+    elif outcome == Outcome.full_handshake:
+        assert to_bytes("SSL_connect:SSLv3/TLS read server certificate") in results.stderr
+    else:
+        assert to_bytes("SSL_accept:error in error") in results.stderr
+
+
+"""
+Basic S2N server happy case.
+
+Tests a single psk connection with no fallback option.
+"""
+@pytest.mark.uncollect_if(func=invalid_test_parameters)
+@pytest.mark.parametrize("cipher", TLS13_CIPHERS, ids=get_parameter_name)
+@pytest.mark.parametrize("curve", ALL_TEST_CURVES, ids=get_parameter_name)
+@pytest.mark.parametrize("protocol", [Protocols.TLS13], ids=get_parameter_name)
+@pytest.mark.parametrize("provider", PSK_PROVIDERS, ids=get_parameter_name)
+@pytest.mark.parametrize("psk_identity", PSK_IDENTITY_LIST, ids=get_parameter_name)
+@pytest.mark.parametrize("psk_secret", PSK_SECRET_LIST, ids=get_parameter_name)
+def test_s2n_server_psk_connection(managed_process, cipher, curve, protocol, provider, psk_identity, psk_secret):
+    port = next(available_ports)
+    random_bytes = data_bytes(10)
+    psk_hash_alg = get_psk_hash_alg_from_cipher(cipher)
+    skip_invalid_psk_tests(provider, psk_hash_alg)
+
+    if provider == S2N:
+        client_psk_params = setup_s2n_psk_params(psk_identity, psk_secret, psk_hash_alg)
+    else:
+        client_psk_params = setup_openssl_psk_params(psk_identity, psk_secret)
```

Comment:
Non-blocking, but if you find yourself implementing something differently for every provider, maybe consider polymorphism? https://github.com/aws/s2n-tls/blob/main/tests/integrationv2/test_early_data.py#L37-L51